src/lib/tuning/abstract_selector.hpp

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "tuning/tuning_choice.hpp"
+#include "tuning/tuning_operation.hpp"
+
+namespace opossum {
+
+/**
+ * An AbstractSelector transforms an unordered list of TuningChoices into
+ * a concrete sequence of TuningOperations.
+ *
+ * It considers a cost budget that the entire operation sequence as well as any
+ * continuous subsequence from the beginning must not exceed.
+ * The operation sequence is prioritized by expected performance impact on the system,
+ * so that the most beneficial operations come before less useful operations.
+ *
+ * For more information on the underlying problem, see e.g. Wikipedia:
+ *   https://en.wikipedia.org/wiki/Knapsack_problem
+ */
+class AbstractSelector {
+ public:
+  virtual ~AbstractSelector() {}
+  /**
+   * Determine the index operation sequence as specified above based on the
+   * given list of choices and the budget value.
+   */
+  virtual std::vector<std::shared_ptr<TuningOperation>> select(
```

Comment:
Select sounds like something is actually selected, but this sounds more like creating an order to me?

Reply:
Yes, this returns an ordered list of *selected* operations. There doesn't have to be a 1:1 mapping of Choices to Operations so it is indeed a selection process. Calling this operation `order` or similar would probably obfuscate this fact. We're open to finding a name that incorporates both aspects :)